<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    بسکتبالیست‌ها
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        در این سوال قصد داریم تا داده‌های مربوط به عملکرد بازیکنان لیگ بسکتبال NBA را تحلیل کنیم.
        پس کافی است تا کار‌هایی که از شما خواسته شده را مرحله به مرحله انجام دهید.
    </font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        مجموعه‌داده
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        مجموعه‌داده‌ای که در احتیار شما قرار گرفته است شامل اطلاعات عملکرد بازیکنان NBA در فصل‌های ۲۰۰۰ الی ۲۰۰۹ است.
    این مجموعه شامل ۱۸۳۰ سطر و ۱۸ ستون می‌باشد که هر سطر آن شامل اطلاعات یک بازیکن در یک فصل است. توضیحات مربوط به ستون‌ها نیز در جدول زیر آمده است.
    </font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>PlAYER</code>|نام بازیکن|
|<code>TEAM</code>|مخفف اسم تیم بازیکن در آن فصل|
|<code>YEAR</code>|سال (فصل مسابقات)|
|<code>GP</code>|تعداد بازی‌های انجام شده|
|<code>MIN</code>|میانگین دقایق بازی کرده|
|<code>PTS</code>|میانگین امتیاز|
|<code>FGM</code>|میانگین تعداد پرتاب‌های ۲ امتیازی موفق|
|<code>FGA</code>|میانگین تعداد کل پرتاب‌های ۲ امتیازی|
|<code>3PM</code>|میانگین تعداد پرتاب‌های ۳ امتیازی موفق|
|<code>3PA</code>|میانگین تعداد کل پرتاب‌های ۳ امتیازی|
|<code>FTM</code>|میانگین تعداد پرتاب‌های آزاد (پنالتی) موفق|
|<code>FTA</code>|میانگین تعداد کل پرتاب‌های آزاد (پنالتی)|
|<code>OREB</code>|میانگین تعداد ریباند‌ها در حمله|
|<code>DREB</code>|میانگین تعداد ریباند‌ها در دفاع|
|<code>AST</code>|میانگین تعداد پاس‌های منجر به امتیاز|
|<code>STL</code>|میانگین تعداد دفعات توپ‌ربایی|
|<code>BLK</code>|میانگین تعداد بلاک‌ها|
|<code>TOV</code>|میانگین تعداد دفعات لو دادن توپ|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
   با استفاده از سلول زیر کتابخانه‌های مورد نیاز خود را <code>import</code> کنید و سپس مجموعه‌داده را که در فایلی به نام <code>NBA2000-2009.csv</code> قرار دارد، بخوانید.
</font>
</p>

In [2]:
import pandas as pd

df = pd.read_csv('NBA2000-2009.csv')
df.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت اول
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         لیستی از بازیکن‌ها و فصل‌ها را بدست آورید که در آن فصل، آن بازیکن در حداقل ۲ تا از ۵ ملاک اصلی آمار دو رقمی (بزرگتر مساوی ۱۰) ثبت کرده باشد.
        <br>
        منظور از ۵ ملاک اصلی میانگین امتیاز، میانگین تعداد ریباند (مجموع ریباند در دفاع و حمله)، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها است.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>double_double</code> قرار دهید.
    این دیتافریم باید شامل ۷ ستون باشد که ستون‌ها از چپ به راست به ترتیب نام بازیکن، سال، میانگین امتیاز، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد ریباند‌ها، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها باشند. همچنین کل دیتافریم باید به ترتیب برحسب سال و نام بازیکن به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|YEAR|PTS|AST|REB|STL|BLK|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Antonio Davis|2000|13.7|1.4|10.1|0.3|1.9|
|Antonio McDyess|2000|20.8|2.1|12.0|0.6|1.5|
    
</font>
</div>
</center>

In [9]:
def check_double_double(row):
  double_digit_counts = sum([row['PTS'] >= 10, row['AST'] >= 10, row['OREB'] >= 10 , row['STL'] >= 10 , row['BLK'] >= 10])
  return double_digit_counts >= 2
double_double = df[df.apply(check_double_double , axis = 1)][0:6]
double_double = double_double.sort_values(by = ['YEAR' , 'PLAYER'])
double_double

PLAYER TEAM  YEAR  GP   MIN   PTS  FGM   FGA  3PM  3PA  FTM  \
219     Andre Miller  CLE  2001  81  37.3  16.5  5.9  12.9  0.3  1.1  4.5   
764       Steve Nash  PHX  2004  75  34.3  15.5  5.7  11.4  1.3  2.9  2.8   
934       Steve Nash  PHX  2005  79  35.4  18.8  6.8  13.4  1.9  4.3  3.3   
1108      Steve Nash  PHX  2006  76  35.3  18.6  6.8  12.8  2.1  4.5  2.9   
1280      Chris Paul  NOH  2007  80  37.6  21.1  7.9  16.1  1.2  3.1  4.2   
1298  Deron Williams  UTA  2007  82  37.3  18.8  6.9  13.6  1.0  2.6  4.0   

      FTA  OREB  DREB   AST  STL  BLK  TOV  
219   5.5   1.3   3.3  10.9  1.6  0.4  3.0  
764   3.2   0.8   2.6  11.5  1.0  0.1  3.3  
934   3.5   0.6   3.6  10.5  0.8  0.2  3.5  
1108  3.3   0.4   3.1  11.6  0.8  0.1  3.8  
1280  4.9   0.8   3.2  11.6  2.7  0.1  2.5  
1298  5.0   0.4   2.5  10.5  1.1  0.3  3.4

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت دوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         باارزش‌ترین بازیکنان این ۱۰ فصل را مشخص کنید.
        ارزش یک بازیکن در یک فصل با استفاده از فرمول زیر بدست می‌آید:
    </font>
    $$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$
    منظور از <code>Missed FG</code> و <code>Missed FT</code> به ترتیب میانگین تعداد پرتاب‌های ۲ امتیازی و آزاد از دست رفته است.  
    <br>
    ارزش یک بازیکن در کل ۱۰ فصل برابر است با میانگین ارزش او در هر فصل.
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>best_player</code> قرار دهید.
    این دیتافریم شامل دو ستون است که یکی نام بازیکن و دومی ارزش آن بازیکن است که تا دو رقم اعشار باید گرد شود. همچنین این دیتافریم باید بر اساس ارزش به صورت نزولی و سپس بر اساس نام بازیکن به‌صورت صعودی مرتب باشد.
    دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|VALUE|
|:------:|:---:|
|Kevin Garnett|29.74|
|LeBron James|27.9|
    
</font>
</div>
</center>

In [14]:
df['Missed FT'] = df['FTA'] - df['FTM']
df['Missed FG'] = df['FGA'] - df['FGM']

df["VALUE"] = df['PTS'] + df['OREB'] + df['DREB'] + df['AST'] + df['STL'] + df['BLK'] - (df['TOV'] + (df['Missed FT'] - df['Missed FG']))
best_player=  df.groupby('PLAYER')['VALUE'].mean().reset_index()
best_player = best_player.sort_values(by = ['VALUE' , 'PLAYER'] , ascending = [False , True])
best_player['VALUE'] = best_player['VALUE'].round(2)
best_player.head()

PLAYER  VALUE
13   Allen Iverson  50.06
335   LeBron James  49.79
323    Kobe Bryant  48.92
537  Tracy McGrady  48.36
99    Chris Webber  47.93

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت سوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         در هر سال چه تیمی بیشترین میانگین امتیاز را ثبت کرده است؟
        <br>
        توجه داشته باشید که میانگین امتیاز یک تیم برابر است با مجموع میانگین امتیاز بازیکن‌هایش.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>max_PTS_of_year</code> قرار دهید.
    این دیتافریم باید شامل سه ستون باشد که ستون‌ها از چپ به راست به ترتیب سال، نام تیم و میانگین امتیاز است که امتیاز‌ها باید تا دو رقم اعشار گرد شوند. همچنین کل دیتافریم باید برحسب سال به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|YEAR|TEAM|PTS|
|:------:|:---:|:---:|
|2000|SAC|100.2|
|2001|LAL|95.9|
    
</font>
</div>
</center>

In [20]:
team_year_points = df.groupby(["YEAR" , "TEAM"])["PTS"].sum().reset_index()
max_PTS_of_year = team_year_points.loc[team_year_points.groupby('YEAR')['PTS'].idxmax()]
max_PTS_of_year['PTS'] = max_PTS_of_year['PTS'].round(2)
max_PTS_of_year = max_PTS_of_year.sort_values(by = ['YEAR'])
max_PTS_of_year.head()

YEAR TEAM    PTS
22   2000  SAC  100.2
41   2001  LAL   95.9
77   2002  ORL   97.0
92   2003  DEN   96.0
117  2004  BOS  106.2

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) در غیر این صورت، در پایان مسابقه نمره شما به صفر تغییر خواهد کرد.
    <br>
    همچنین اگر از کولب برای اجرای این فایل نوت‌بوک استفاده می‌کنید، قبل از ارسال فایل <code>result.zip</code>، آخرین نسخه‌ی نوت‌بوک خود را دانلود کرده و داخل فایل ارسالی قرار دهید.
</font>

In [21]:
import zipfile
import os
if not os.path.exists(os.path.join(os.getcwd(), 'NBA.ipynb')):
    %notebook -e NBA.ipynb

double_double.to_csv('double_double.csv', index=False)
best_player.to_csv('best_player.csv', index=False)
max_PTS_of_year.to_csv('max_PTS_of_year.csv', index=False)

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)

file_names = ['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
compress(file_names)

File Paths:
['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
